polygeist/lib/polygeist/Passes/ConvertToLLVMABI.cpp

```diff
@@ -178,6 +178,111 @@ class SYCLConstructorLowering
   }
 };
 
+static mlir::Value castToBaseType(PatternRewriter &Rewriter, mlir::Location Loc,
+                                  mlir::Value Original,
+                                  mlir::MemRefType BaseType) {
+  const auto ThisType = Original.getType().cast<MemRefType>();
+
+  const auto TargetShape = BaseType.getShape();
+  const auto TargetElementType = BaseType.getElementType();
+  const auto TargetMemSpace = BaseType.getMemorySpaceAsInt();
+
+  assert(TargetShape == ThisType.getShape() &&
+         "Shape should not change when casting to base class for a member "
+         "function call.");
+
+  // The element type will always change here.
+  mlir::Value Cast = Rewriter.create<sycl::SYCLCastOp>(
+      Loc,
+      MemRefType::get(TargetShape, TargetElementType, {},
+                      ThisType.getMemorySpaceAsInt()),
+      Original);
+
+  LLVM_DEBUG(llvm::dbgs() << "  Cast inserted: " << Cast << "\n");
+
+  const auto CurrMemorySpace = ThisType.getMemorySpaceAsInt();
+
+  // Cast address space if needed too.
+  if (TargetMemSpace != CurrMemorySpace) {
```

Comment:
Should we assert this is not the case as in `createSYCLMethodOp` we ensure the memory space is not changed.
```
// Make sure the memory space is not changed:
    const auto MemSpace =
        operandsCpy[0].getType().cast<MemRefType>().getMemorySpaceAsInt();
    if (NewArg.getType().cast<MemRefType>().getMemorySpaceAsInt() != MemSpace) {
      NewArg = castToMemSpace(NewArg, MemSpace);
    }
```

Reply:
True. Asserted